docs/source/concepts/InstrumentAccessLayers.rst

```diff
@@ -31,216 +33,170 @@ ____________
+The following methods are useful helpers on ``ComponentInfo`` that allow the extraction of the **component index** for key components
 
-**Before refactoring**
-
-.. code-block:: c++
+* ``root()``
+* ``source()``
+* ``sample()``
 
-  try {
-    Geometry::IDetector_const_sptr det = ws.getDetector(wsIndex);
-    // do stuff
-  catch (Exception::NotFoundError &) {
-    // do exception
-    return false;
-  }
-
-**After - check it has some detectors**
-
-.. code-block:: c++
+**Indexing**
 
-  #include "MantidAPI/SpectrumInfo.h"
+The ``ComponentInfo`` object is accessed by an index going from 0 to the number of components (including the instrument iteself). **ALL Detector Indexes have a Component Index which is the EXACT same representation**, this is an important point to understand. In other words, a detector with a Detector Index of 5, for the purposes of working with a ``DetectorInfo`` and  will have a Component Index of 5, when working with a ``ComponentInfo``. Explained in yet another way: The first 0 - n components referenced in the ``ComponentInfo`` are detectors, where n is the total number of detectors. This gurantee can be leveraged to provide speedups, as some of the examples will show.  
 
-  ...
+A ``ComponentID`` for compatiblity with older code, and be extracted from ``ComponentInfo::componentID(componentIndex)``, but such calls should be avoided where possible.
 
-  const auto &spectrumInfo = ws->spectrumInfo();
+It is also possible to use the method ``componentInfo.indexOf(componentID)`` to get the index for a particular component ID. However, this is a call to a lookup in an unordered map, so is an expensive calculation which should be avoided where possible.
 
-  if (spectrumInfo.hasDetectors(wsIndex)) {
-    // do stuff
-  } else {
-    // do exception
-    return false;
-  }
+**One should NEVER expose a Component Index or Detector Index through a public facing (python, gui, ...) to an end user**. Detector Index and Component Indexes are internal concepts for fast enumeration. It is however desirable to translate from a ``ComponentIndex`` via ``ComponentInfo::indexOf`` to as early as possible and in such a way to avoid repeated calls to this method, as stated above. Likewise, conversion back to a ``ComponentIndex``, if so required, should be done as infrequently and, as late as possible.
 
-In this case, which is generally more common, the check is for at least one detector. It is also possible to check for the existence of a unique detector, see the alternative after example below.
+Below is an example refactoring.
 
-**After - check for a unique detector**
+**Before refactoring**
 
 .. code-block:: c++
 
-  #include "MantidAPI/SpectrumInfo.h"
-
-  ...
-
-  const auto &spectrumInfo = ws->spectrumInfo();
+  auto instrument = ws->getInstrument();
+  if (!instrument)
+    throw runtime_error("There is no instrument in input workspace.")
```

Comment:
A workspace always has an instrument, so the check in the old code is pointless.